Wrappers/Python/cil/optimisation/functions/TotalVariation.py

```diff
@@ -199,6 +212,24 @@ def __init__(self,
         else:
             self.func = MixedL11Norm()
 
+     # setting the dual value for the proximal calculation - important for warmstart
+    @property
+    def p2(self):
+        if self._p2 is not None:
+            return self._p2
+        else:
+            p2 = self.gradient.range_geometry().allocate(0)
+            if self.warmstart:
+                self.p2 = p2
+            return p2
+
+    @p2.setter
+    def p2(self, value):
+        if isinstance(value, type(self.gradient.range_geometry().allocate(0))):
```

Comment:
This test would allocate an the range of the gradient just to test the class type. I don't think it's wise.
`range_geometry` will always return a `BlockGeometry`
https://github.com/TomographicImaging/CIL/blob/a2e45fe0aafe6329ec9027bda1cf262d4aff0284/Wrappers/Python/cil/optimisation/operators/GradientOperator.py#L360-L362 and https://github.com/TomographicImaging/CIL/blob/a2e45fe0aafe6329ec9027bda1cf262d4aff0284/Wrappers/Python/cil/optimisation/operators/GradientOperator.py#L222

`BlockGeometry` does not have an `__equals__` operator so we would need to construct it. Maybe it could be sufficient to check the dimensionality of the data passed, i.e.

```suggestion
        if len(value.geometry.geometries) != len(self.gradient.range_geometry().geometries):
``` 

With a 2D (3D, 4D) dataset we expect the range of the gradient to be a `BlockGeometry` with 2 (3, 4) geometries inside.

Reply:
Is now 

     if len(value) != len(self.gradient.range_geometry().geometries):
            self._p2 = value

Reply:
We have since removed the setter